### ISSUE TITLE
The script to download cc_matrix is not usable

### ISSUE THREAD

**Issue by Morizeyao** [facebookresearch/LASER#156]
In this file

(https://github.com/facebookresearch/LASER/blob/master/tasks/CCMatrix/dl_cc_matrix.py)
line 9 (from cc_net.process_wet_file import CCSegmentsReader)

The code raises an error.

It seems there is no CCSegmentsReader to import.
(https://github.com/facebookresearch/cc_net/blob/master/cc_net/process_wet_file.py)

Am I doing something wrong?

**Comment by mehmedes** [comment]
Using https://github.com/facebookresearch/cc_net/blob/dev/cc_net/jsonql.py and https://github.com/facebookresearch/cc_net/blob/dev/cc_net/process_wet_file.py from the LASER's DEVbranch should do the trick

**Comment by gwenzek** [comment]
Hi, sorry for the current mess.
I need to merge the dev branch of cc_net and make a release,
but there was some testing I wanted to do before that.

The data isn't ready yet anyway, but we are working hard on it.
I'll update the issue as I make progress.

**Comment by gwenzek** [comment]
We have a beta version of the data ready
We are still missing some pairs including the ones with English, Japanese and Chinese.
@hoschwenk  is working on that.

You can try downloading them by using 835c0708a285a667efd7f8316fda67877a2f4283 and

```sh 
pip install cc_net func_argparse
python tasks/CCMatrix/dl_cc_matrix.py dl --help
python tasks/CCMatrix/dl_cc_matrix.py dl -v v1.0.beta
```

Downloading is using 8 processes in parallel, you can increase that if your network isn't maxed out.
Once this is over you'll need `python tasks/CCMatrix/dl_cc_matrix.py finalize -v v1.0.beta` to extract the bitext.

**Comment by mehmedes** [comment]
Thank you!!! @gwenzek @hoschwenk

**Comment by thpun** [comment]
@gwenzek
In dl_cc_matrix.py, there is `v1.0.alpha` & `v1.0.beta`. What's the difference of two?

**Comment by gwenzek** [comment]
> In dl_cc_matrix.py, there is v1.0.alpha & v1.0.beta. What's the difference of two?

v1.0.alpha contains 419 language pairs 
v1.0.beta contains 878 language pairs and is a superset of v1.0.alpha.

I'm reserving v1.0 for once we have all the data.

**Comment by tmkhalil** [comment]
@gwenzek 
I'm trying to download the data but it feels like it's taking forever, what is the expected size of the "raw" directory after downloading "v1.0.beta"?
Thank you!

**Comment by gwenzek** [comment]
It should be 641 Gb big and will download over 64 Tb of data to look for the bitext. So yes it'll probably take a long time.
We are not allowed to distribute a subset of commoncrawl, that's why we have to use this overcomplicated process to allow people to reproduce our work.

**Comment by jorgtied** [comment]
After about 1 day and running with 16 threads I am at about 4Gb (the compressed data in data/raw). Is that normal? Extrapolating from that it would take me 160 days to fetch the data. I would really like to speed that up somehow. Any hint? My bandwidth for downloading should be OK and most of the threads are busy 100%.

**Comment by gwenzek** [comment]
Can you try with more threads ? You can have more threads than CPU core because you want to maximize the network usage.
You probably also want to run this from a server with a good download speed.
And probably in California, I'm not sure where the data is served from.

**Comment by jorgtied** [comment]
OK - I will try that. Download speed should be pretty good. I get info about those rates - does that look reasonable?
```
2020-12-01 17:44 INFO 44165:root - Downloaded https://commoncrawl.s3.amazonaws.com/crawl-data/CC-MAIN-2018-43/segments/1539583508988.18/wet/CC-MAIN-20181015080248-20181015101748-00392.warc.wet.gz [200] took 9s (18699.7kB/s)
2020-12-01 17:44 INFO 44178:root - Downloaded https://commoncrawl.s3.amazonaws.com/crawl-data/CC-MAIN-2018-17/segments/1524125945497.22/wet/CC-MAIN-20180422061121-20180422081121-00638.warc.wet.gz [200] took 9s (16614.0kB/s)
2020-12-01 17:44 INFO 44179:cc_net.process_wet_file - Kept 53_242 documents over 53_242 (100.0%).
2020-12-01 17:44 INFO 44175:root - Downloaded https://commoncrawl.s3.amazonaws.com/crawl-data/CC-MAIN-2018-13/segments/1521257645248.22/wet/CC-MAIN-20180317155348-20180317175348-00391.warc.wet.gz [200] took 6s (19065.1kB/s)
2020-12-01 17:44 INFO 44179:root - Starting download of https://commoncrawl.s3.amazonaws.com/crawl-data/CC-MAIN-2018-43/segments/1539583512434.71/wet/CC-MAIN-20181019191802-20181019213302-00125.warc.wet.gz
2020-12-01 17:44 INFO 44174:cc_net.process_wet_file - Kept 39_203 documents over 39_203 (100.0%).
```

**Comment by bmtm** [comment]
the commoncrawl data is served by aws, from the us-east-1 region. I'm trying to run the script on a 32 thread system in the same az

am I correct in assuming that 64Tb should be downloaded, resulting in 641Gb on disk? I've downloaded ~2Tb so far but there is only ~600Mb on disk, which is more than an order of magnitude off. Do you mean 641Gb after decompressing, or should the .tsv.gz files add up to 641Gb?

**Comment by bbss** [comment]
I am echoing what @bmtm is saying/asking, I've downloaded nearly 64TB so far, and all the files in the raw directory only add up to about 24GB

**Comment by jorgtied** [comment]
Yet another question: is it possible to resume an already started download process in case it crashes on the way? I started my process on a cluster with a 72 hours walltime but that will not be enough. Do I have to start from scratch?

**Comment by bmtm** [comment]
@jorgtied if my reading of the code is correct, any completed .tsv.gz files are skipped (but only if they have been completed, without a .tmp suffix. The files ending with .tmp.tsv.gz will need to be re-downloaded)

**Comment by bmtm** [comment]
I'm still downloading, but noticed a few possible (?) issues with the code:
- found_bitext should probably be incremented after writing the bitext to file (just a logging issue, shouldn't affect the download)
- in the dl function, o.close() is called without a success variable, so the .tmp files are never renamed. This again doesn't affect the download but would prevent restarting since there's no way to tell which files are tmp and which are completed
- I've logged an uncaught exception that prematurely ended some of the threads:
```
2020-12-03 11:54 ERROR 2411:root - Traceback (most recent call last):
  File "dl_cc_matrix.py", line 139, in dl_file
    raw_documents = get_documents(segment)
  File "dl_cc_matrix.py", line 107, in get_documents
    return {d["digest"]: d["raw_content"] for d in CCSegmentsReader([segment])}
  File "dl_cc_matrix.py", line 107, in <dictcomp>
    return {d["digest"]: d["raw_content"] for d in CCSegmentsReader([segment])}
  File "/home/ubuntu/.local/lib/python3.8/site-packages/cc_net/process_wet_file.py", line 199, in __iter__
    for doc in parse_warc_file(self.open_segment(segment), self.min_len):
  File "/home/ubuntu/.local/lib/python3.8/site-packages/cc_net/process_wet_file.py", line 132, in parse_warc_file
    for doc in group_by_docs(lines):
  File "/home/ubuntu/.local/lib/python3.8/site-packages/cc_net/process_wet_file.py", line 105, in group_by_docs
    for warc in warc_lines:
  File "/home/ubuntu/.local/lib/python3.8/site-packages/cc_net/jsonql.py", line 971, in _close_when_exhausted
    yield from file
  File "/usr/lib/python3.8/gzip.py", line 299, in read1
    return self._buffer.read1(size)
  File "/usr/lib/python3.8/_compression.py", line 68, in readinto
    data = self.read(len(byte_view))
  File "/usr/lib/python3.8/gzip.py", line 492, in read
    raise EOFError("Compressed file ended before the "
EOFError: Compressed file ended before the end-of-stream marker was reached
```
this does not appear to account for the low size of files on disk, however. If the .tsv.gz files are uniform in size I will end up with 58GB on disk, which seems too small?

**Comment by jorgtied** [comment]
I still have big problems downloading the dataset. When trying to increase the number of treads to 32, 64 or even 128 my process gets stalled and does not continue. I get ssl connection errors and it seem that I hit the wall of download limits. 

UserWarning: Swallowed error HTTPSConnectionPool(host='dl.fbaipublicfiles.com', port=443): Max retries exceeded with url: /laser/CCMatrix/v1.0.beta/2019-51_0036.tsv.gz (Caused by SSLError(SSLError(1, '[SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:2607)'))) while downloading https://dl.fbaipublicfiles.com/laser/CCMatrix/v1.0.beta/2019-51_0036.tsv.gz (1 out of 3)

I don't really know how to proceed. What would be a reasonable way of obtaining the data?

**Comment by tmkhalil** [comment]
@gwenzek Is it possible to allow specifying the language pair? This can make it possible to download a pair in a reasonable time if you don't want the rest

**Comment by gwenzek** [comment]
I'll try to clarify a bit of what's going on.
For legal reasons we can't release directly the bitext. Instead we release pointers to the bitext.
For each bitext we give you the segment, the document digest and the string offset into the document.
Segments are 64Mb .gz files containing the text, distributed by [CommonCrawl](https://commoncrawl.org/).

We have indexed the equivalent of 64Tb compressed text from CC, ie ~256 Tb of uncompressed text.
The sentences from CC matrix are spread over those 256 Tb of raw text

To avoid people to have to __store__ all of this (but you'll still need to __download__), I released the sentences sorted by segments.
The pipeline in `dl_cc_matrix.py` is doing the following:

1. download the bitext pointers sorted by segment, XXGb split in 32 files
2. read through all those files, for each segment:
   a. download it and decompress it to memory
   b. extract all the sentences we are interested in from this segment
   c. discard the segment
3. split the extracted segment by language pairs
4. for each language pair sort the source sentences by bitext id, and target sentences by bitext id.
5. zip the two source sentences and target sentences files together and check that the bitext ids match.

If you have better ideas on how to release this data, I'll be happy to change the release format and republish the data, but I'll only ever be allowed to release text pointers not actual text.

If you want to make this more robust to interruption:
* add checkpointing in step 2. I'll look into this this week.

If you want to make this faster:
* filter by language pair __before__ downloading the segment. This helps only if you restrict to a very small subset of the language pairs, because otherwise you'll still need to download/decompress most of the segments.

For reference the job for preparing the final release has been running for two weeks on 256 workers. I expect `dl_cc_matrix.py` to take a similar amount of time and workers. Segments files where downloaded at around ~30Mb/s.

Of course this is Python code and we could seem some gains by using eg a faster json parser, but just decompression, download, regex takes >50% of time in my experience, so that puts an upper bound of speed improvement at 2x.

Also I'd like to point out that we haven't finished releasing the CC-matrix dataset and you're playing with an intermediary version. I'm glad to have beta tester to help me iron this out, but be aware than their are probably more bugs laying ahead.

**Comment by Jack000** [comment]
@gwenzek thanks for explaining.

How big should each final .tsv.gz file be though? It's 88Mb for me, so with 664 files it will come to 58GB - does that sound about right?

**Comment by bmtm** [comment]
so my download has completed, and the raw dir contains 659 .tsv.gz files each about 87-91MB, for a total of 57GB.

I tried running the finalize step but ran into some issues:
- lang_pairs are formatted as "en-de/de" the / is interpreted as a part of a file path, but the en-de directory is never created. This happens in both transpose_file() and sort_files()
- raw_dir.iterdir() if d.is_dir() should probably be split_dir.iterdir() (?) as there are no directories in the raw dir
- got this exception:
multiprocessing.pool.RemoteTraceback: 
```
"""
Traceback (most recent call last):
  File "/home/jack/miniconda3/lib/python3.8/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/home/jack/miniconda3/lib/python3.8/multiprocessing/pool.py", line 48, in mapstar
    return list(map(*args))
  File "dl_cc_matrix.py", line 242, in transpose_file
    o = FileWriterWithTmp(outdir / f"{lang_pair}_{file.name}")
  File "dl_cc_matrix.py", line 189, in __init__
    self.fd, self.tmp_file = _tmp(file)
  File "dl_cc_matrix.py", line 182, in _tmp
    fd, tmp_path = tempfile.mkstemp(dir=tmp_dir, prefix=prefix, suffix=suffix)
  File "/home/jack/miniconda3/lib/python3.8/tempfile.py", line 332, in mkstemp
    return _mkstemp_inner(dir, prefix, suffix, flags, output_type)
  File "/home/jack/miniconda3/lib/python3.8/tempfile.py", line 250, in _mkstemp_inner
    fd = _os.open(file, flags, 0o600)
OSError: [Errno 24] Too many open files: 'data/v1.0.beta/split_by_lang/ja-ko/ko_2017-47_0002.2pnc37sg.tmp.kitmpukv.tmp.tsv.gz'
"""
```
presumably a file handle is not closed somewhere, but I'm not certain where. I just worked around it by restarting the script on a loop.

- in sort_files(), the sort command operates on a list of .gz files. I think there was meant to be an extraction step here, because sort only works on raw text as far as I know.

after this, I managed to get 70GB in the bitext directory but a lot of the bitext directories are empty. I'm guessing something went wrong but I'm not sure where exactly. Anyways, I think I'll wait for the official release which will hopefully address these issues.

**Comment by gwenzek** [comment]
Each .tsv.gz files will have very various sizes.
Sentences mostly come from the "first" snapshots, and some snapshots have very little content.
The final size should be around 500Gb.

I've looked at the content of the AWS bucket and it was missing the biggest files. Sorry for not having spotted this earlier. I've launched the upload but it's still ongoing. I'll let you know once the files of the beta version are all there.

**Comment by jorgtied** [comment]
Just wonder whether the files are now there and I can start a new attempt to download the whole thing ... Thanks!

**Comment by sukuya** [comment]
@gwenzek  Any update on this.

**Comment by RPrenger** [comment]
> so my download has completed, and the raw dir contains 659 .tsv.gz files each about 87-91MB, for a total of 57GB.
> 
> I tried running the finalize step but ran into some issues:
> 
> * lang_pairs are formatted as "en-de/de" the / is interpreted as a part of a file path, but the en-de directory is never created. This happens in both transpose_file() and sort_files()
> * raw_dir.iterdir() if d.is_dir() should probably be split_dir.iterdir() (?) as there are no directories in the raw dir
> * got this exception:
>   multiprocessing.pool.RemoteTraceback:
> 
> ```
> """
> Traceback (most recent call last):
>   File "/home/jack/miniconda3/lib/python3.8/multiprocessing/pool.py", line 125, in worker
>     result = (True, func(*args, **kwds))
>   File "/home/jack/miniconda3/lib/python3.8/multiprocessing/pool.py", line 48, in mapstar
>     return list(map(*args))
>   File "dl_cc_matrix.py", line 242, in transpose_file
>     o = FileWriterWithTmp(outdir / f"{lang_pair}_{file.name}")
>   File "dl_cc_matrix.py", line 189, in __init__
>     self.fd, self.tmp_file = _tmp(file)
>   File "dl_cc_matrix.py", line 182, in _tmp
>     fd, tmp_path = tempfile.mkstemp(dir=tmp_dir, prefix=prefix, suffix=suffix)
>   File "/home/jack/miniconda3/lib/python3.8/tempfile.py", line 332, in mkstemp
>     return _mkstemp_inner(dir, prefix, suffix, flags, output_type)
>   File "/home/jack/miniconda3/lib/python3.8/tempfile.py", line 250, in _mkstemp_inner
>     fd = _os.open(file, flags, 0o600)
> OSError: [Errno 24] Too many open files: 'data/v1.0.beta/split_by_lang/ja-ko/ko_2017-47_0002.2pnc37sg.tmp.kitmpukv.tmp.tsv.gz'
> """
> ```
> 
> presumably a file handle is not closed somewhere, but I'm not certain where. I just worked around it by restarting the script on a loop.
> 
> * in sort_files(), the sort command operates on a list of .gz files. I think there was meant to be an extraction step here, because sort only works on raw text as far as I know.
> 
> after this, I managed to get 70GB in the bitext directory but a lot of the bitext directories are empty. I'm guessing something went wrong but I'm not sure where exactly. Anyways, I think I'll wait for the official release which will hopefully address these issues.

@bmtm 
I ran into the same issues.  But I believe the interpretation of the language pair encodings as directories may be intentional, because the finalize script sort of assumes the directories are laid out that way.  I needed to create the directories and change the raw_dir variables to split_dir as well like you mentioned

I also found that the code seems to be calling the program `sort` on the .gz files directly without unzipping them.  I did a PR with my fixes here:
https://github.com/facebookresearch/LASER/pull/172

Finally I found that several of the text segments have newlines, but the code is assuming records are split on newlines.  So I ran a script that replaces those newlines with spaces in the `raw` directory before calling finalize, but I haven't added that code to the PR yet, because I assume I'm missing something and it shouldn't be necessary.

**Comment by jorgtied** [comment]
Is there some update here? Now I have 663 files and after some struggling with the fixes proposed above I managed to start the finalize step but all bitexts so far are invalid (different number of lines in source and target and they do not look like translations). Is this because some files are missing from the download? Did anyone manage to actually download and prepare the data? Any help is appreciated!

**Comment by jorgtied** [comment]
@RPrenger - did you get any reasonable bitext segments from the download script? For me all of them are invalid with different numbers of lines in source and target text files. I tried to fix all the problems with newline characters, sorting of gzipped files etc. Is it because of the missing files (@gwenzek)?

**Comment by EL-SHREIF** [comment]
I Tried to run the script and I get this error what could be the fix for it?
`2021-01-27 11:37 INFO 12880:root - Downloaded https://commoncrawl.s3.amazonaws.com/crawl-data/CC-MAIN-2018-22/segments/1526794867559.54/wet/CC-MAIN-20180526151207-20180526171207-00494.warc.wet.gz [200] took 2s (67331.2kB/s)                 multiprocessing.pool.RemoteTraceback:                                                                                   """                                                                                                                     Traceback (most recent call last):                                                                                        File "C:\Users\aelsherif\AppData\Local\Programs\Python\Python39\lib\multiprocessing\pool.py", line 125, in worker         result = (True, func(*args, **kwds))                                                                                  File "C:\Users\aelsherif\AppData\Local\Programs\Python\Python39\lib\multiprocessing\pool.py", line 48, in mapstar         return list(map(*args))                                                                                               File "B:\LASER\tasks\CCMatrix\dl_cc_matrix.py", line 137, in dl_file                                                      raw_documents = get_documents(segment)                                                                                File "B:\LASER\tasks\CCMatrix\dl_cc_matrix.py", line 106, in get_documents                                                return {d["digest"]: d["raw_content"] for d in CCSegmentsReader([segment])}                                           File "B:\LASER\tasks\CCMatrix\dl_cc_matrix.py", line 106, in <dictcomp>                                                   return {d["digest"]: d["raw_content"] for d in CCSegmentsReader([segment])}                                           File "C:\Users\aelsherif\AppData\Local\Programs\Python\Python39\lib\site-packages\cc_net\process_wet_file.py", line 199, in __iter__                                                                                                              for doc in parse_warc_file(self.open_segment(segment), self.min_len):                                                 File "C:\Users\aelsherif\AppData\Local\Programs\Python\Python39\lib\site-packages\cc_net\process_wet_file.py", line 132, in parse_warc_file                                                                                                       for doc in group_by_docs(lines):                                                                                      File "C:\Users\aelsherif\AppData\Local\Programs\Python\Python39\lib\site-packages\cc_net\process_wet_file.py", line 105, in group_by_docs                                                                                                         for warc in warc_lines:                                                                                               File "C:\Users\aelsherif\AppData\Local\Programs\Python\Python39\lib\site-packages\cc_net\jsonql.py", line 971, in _close_when_exhausted                                                                                                           yield from file                                                                                                       File "C:\Users\aelsherif\AppData\Local\Programs\Python\Python39\lib\encodings\cp1252.py", line 23, in decode              return codecs.charmap_decode(input,self.errors,decoding_table)[0]                                                   UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 380: character maps to <undefined>               """                                                                                                                                                                                                                                             The above exception was the direct cause of the following exception:                                                                                                                                                                            Traceback (most recent call last):                                                                                        File "B:\LASER\tasks\CCMatrix\dl_cc_matrix.py", line 327, in <module>                                                     func_argparse.main(dl, finalize)                                                                                      File "C:\Users\aelsherif\AppData\Local\Programs\Python\Python39\lib\site-packages\func_argparse\__init__.py", line 29, in main                                                                                                                    return make_main(*fns, module=module, description=description)(sys.argv[1:])                                          File "C:\Users\aelsherif\AppData\Local\Programs\Python\Python39\lib\site-packages\func_argparse\__init__.py", line 72, in parse_and_call                                                                                                          return command(**parsed_args)                                                                                         File "B:\LASER\tasks\CCMatrix\dl_cc_matrix.py", line 102, in dl                                                           pool.map(dlf, file_list)                                                                                              File "C:\Users\aelsherif\AppData\Local\Programs\Python\Python39\lib\multiprocessing\pool.py", line 364, in map            return self._map_async(func, iterable, mapstar, chunksize).get()                                                      File "C:\Users\aelsherif\AppData\Local\Programs\Python\Python39\lib\multiprocessing\pool.py", line 771, in get            raise self._value                                                                                                   UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 380: character maps to <undefined>`

**Comment by EL-SHREIF** [comment]
I don't know if it's okay or not but most of the files that the first step download like "2019-51_0000.v82rc62k.tmp.tsv.gz" has another file with the same name but with the extension ".index" like "2019-51_0000.v82rc62k.tmp.tsv.gz.index"? Is this okay? @gwenzek

**Comment by RPrenger** [comment]
@jorgtied I did eventually get some language pairs that looked reasonable.  However I don't see any data with English though since I've last tried (that was going to be released later).  Sorry about the delay response.  The CC_Aligned data set was released and it's much easier to work with so I've been focusing on that for now.  I'd be interested to hear if there's a new version of CCMatrix with all the language pairs and finalize code.

**Comment by gwenzek** [comment]
Hi all, we had some issues with the release of the missing dataset. I wasn't able to find the bitext for the following languages:
am, ast, cy, fy, ga, gd, ha, ig, jv, km, lb, lg, my, om, or, so, wo, xh, yi, yo, zu. 

I'll still continue trying to recover them but in the meantime I added v1.0.0 on S3 which should have the rest of the languages.
My validation script hasn't finished running , so I hope I'm not creating false hope here. But since I'm away next week and the data take a long time to download anyway, I figured I should share it.

Thanks for your patience and sorry for the delay.

**Comment by jorgtied** [comment]
Thanks for the updates and releasing the v1.0.0.
The download script is now running for about 2 weeks and I wonder what I should expect as the final result to estimate how long the whole process will take. In the raw data directory there is now 265GB of data files. What would be the final size that I should receive there? That would help me to make a guess ...

**Comment by EL-SHREIF** [comment]
Thanks for providing the download scripts to get the data.
I have explored the data and I found some errors like for example: 
Take a look at **"de-en"** specifically:
in **en.txt** we have:
735     1.226383        The answer is found in God’s word to Moses and Israel.
735     1.2264837       The long-awaited political agreement in Zimbabwe is to be welcomed.
735     1.2264837       The long-awaited political agreement in Zimbabwe is to be welcomed.
in **de.txt** we have:
735     1.2264837       Das lang erwartete politische Abkommen in Simbabwe sollte begrüßt werden.
742     1.226383        Die Antwort findet sich im Wort Gottes an Mose und Israel.
We have around **55M** line with duplicate id in the en.txt file out of **288M** line total.
I just need to know what does it mean when we have multiple lines with the same sentence id in the bitext files?
